cpp/src/aztec3/circuits/rollup/merge/.test.cpp

```diff
@@ -9,41 +9,50 @@ using aztec3::circuits::rollup::native_merge_rollup::BaseOrMergeRollupPublicInpu
 using aztec3::circuits::rollup::native_merge_rollup::merge_rollup_circuit;
 using aztec3::circuits::rollup::native_merge_rollup::MergeRollupInputs;
 using aztec3::circuits::rollup::native_merge_rollup::NT;
+using DummyComposer = aztec3::utils::DummyComposer;
 
 } // namespace
 namespace aztec3::circuits::rollup::merge::native_merge_rollup_circuit {
 
 class merge_rollup_tests : public ::testing::Test {};
 
-#ifndef __wasm__ // TODO: temporary hack. WASM doesn't support EXPECT_DEATH or try/catch or EXPECT_THROW so no way to
-                 // test these in wasm.
 TEST_F(merge_rollup_tests, test_different_rollup_type_fails)
 {
+    DummyComposer composer = DummyComposer();
     auto mergeInput = dummy_merge_rollup_inputs_with_vk_proof();
     mergeInput.previous_rollup_data[0].base_or_merge_rollup_public_inputs.rollup_type = 0;
     mergeInput.previous_rollup_data[1].base_or_merge_rollup_public_inputs.rollup_type = 1;
-    EXPECT_DEATH(merge_rollup_circuit(mergeInput), ".*assert_both_input_proofs_of_same_rollup_type.*");
+    merge_rollup_circuit(composer, mergeInput);
+    ASSERT_TRUE(composer.has_failed());
+    ASSERT_EQ(composer.get_first_failure(), "input proofs are of different rollup types");
 }
 
 TEST_F(merge_rollup_tests, test_different_rollup_height_fails)
 {
+    DummyComposer composer = DummyComposer();
     auto mergeInput = dummy_merge_rollup_inputs_with_vk_proof();
     mergeInput.previous_rollup_data[0].base_or_merge_rollup_public_inputs.rollup_subtree_height = 0;
     mergeInput.previous_rollup_data[1].base_or_merge_rollup_public_inputs.rollup_subtree_height = 1;
-    EXPECT_DEATH(merge_rollup_circuit(mergeInput), ".*assert_both_input_proofs_of_same_height_and_return.*");
+    merge_rollup_circuit(composer, mergeInput);
+    ASSERT_TRUE(composer.has_failed());
+    ASSERT_EQ(composer.get_first_failure(), "input proofs are of different rollup heights");
```

Comment:
Do these asserts work in wasm?

Reply:
I'm not sure. This was the approach for Aztec Connect, so I'd hope so. But might be worth quickly running a failing test through wasm, to test whether it's caught.

Reply:
tested [here](https://github.com/AztecProtocol/aztec3-circuits/pull/212)

Seems like wasm is catching failures - https://app.circleci.com/pipelines/github/AztecProtocol/aztec3-circuits/822/workflows/d85743cd-455e-46c3-a300-9108677f7347/jobs/4200

Reply:
It's fixing `asserts`, which don't get caught by wasm.
So instead of asserts, we're finding failures using this new 'composer' class (similar to how bberg circuit composers find failures).
The wasm gtests should now be able to test for failures in wasm

Reply:
Ah I was confused... Thank for the explanation Mike!